org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java

```diff
@@ -58,6 +60,26 @@ public static AbstractUIPlugin getDefault() {
 		return singleton;
 	}
 
+	public static int getDeviceZoom() {
```

Comment:
This implementation will probably suffer from the same problem that `DPIUtil` was suffering which is: the value will be obtained once and never updated again _i.e._ if the zoom of the monitor changes then this value will be outdated.

Reply:
How was this solved in SWT?

Reply:
`DPIUtil.setDeviceZoom` is called on zoom changes, which keeps the `deviceZoom` always up-to-date _i.e._ calling `DPIUtil.getDeviceZoom()` should always give you the current zoom. But those methods are not accessible in this case, so this is just fyi.

In this specific case, you could simply avoid storing the zoom in the field `deviceZoom` and query the system property `org.eclipse.swt.internal.deviceZoom` every time. Since `DPIUtil.setDeviceZoom(...)` is **also** updating that property then the result should be that querying it always gives you the right zoom :-)

Long story short: remove the field `deviceZoom` and remove the condition `if (deviceZoom == -1)`.

Reply:
That you for the deeper insight in how this is handled on the SWT side of things. This is pretty similar to what I had in mind.

Though there is a separate issue which makes me concerned about how much we can get out of it in the first place. If the zoom changes, we normally would also need to refresh the cursors, as they are still using the old image data.

But those cursors are API, meaning we can't simply dispose them and create new objects...

Reply:
With the latest state, the device zoom is no longer stored in a static variable, but always read directly from the system property.

Reply:
> That you for the deeper insight in how this is handled on the SWT side of things. This is pretty similar to what I had in mind.
> 
> Though there is a separate issue which makes me concerned about how much we can get out of it in the first place. If the zoom changes, we normally would also need to refresh the cursors, as they are still using the old image data.
> 
> But those cursors are API, meaning we can't simply dispose them and create new objects...

Yes, that is true. The problems are that:
1. The cursors are **constants** in `SharedCursors` _i.e._ one can not simply recreate them when the zoom changes
2. `Image` and `ImageData` don't really know anything about zoom changes which means once you create a `Cursor` out of an `Image`/`ImageData`, you're stuck with the zoom level at which they were created.

Point nr. 2 means that one has to manually react to zoom level changes and recreate the cursors with an image/image data at the appropriate zoom level, but this breaks the contract established in point nr. 1.

Sadly, I don't see an easy way out of this problem, at least not at this level of abstraction (GEF is a consumer of SWT). Best case scenario, this kind of issues would be handled transparently by SWT and not by the consumer (GEF), but I know this is not trivial since our team has been working on the [HighDPI subject](https://github.com/orgs/vi-eclipse/projects/1/views/1?filterQuery=label%3AHiDPI) for quite a while now and this issue has also surfaced [somewhere else](https://github.com/vi-eclipse/Eclipse-Platform/issues/65).

I will take this to the team and we'll try to address it in https://github.com/vi-eclipse/Eclipse-Platform/issues/65 too.

Reply:
> Point nr. 2 means that one has to manually react to zoom level changes and recreate the cursors with an image/image data at the appropriate zoom level, but this breaks the contract established in point nr. 1.

That's what I'm afraid of. And it's not just affecting the SharedCursors class, but also the Draw2D Cursors class containing static instances of the native SWT cursors. In the long term, I believe the most sensible thing to do would be to deprecate those static fields and have something similar to the ImageDataProvider, except for cursors. 

GEF can then listen to SWT.ZoomChanged events and reload the cursors accordingly. But that is going way beyond the scope of this PR and something we have to consider at a later date. Thank you again for your input.

Reply:
> Which class is that? If it's using the native cursors (i.e. the constructor that takes a style as parameter) then they shouldn't be affected by this issue.

I'm talking about this [Cursors](https://github.com/eclipse/gef-classic/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/Cursors.java) class, where we create static instances of the native cursors. Looking at the actual implementation, it looks like they don't actually work with ImageData instances, so you are indeed correct and they should be updated automatically.